python/ray/dag/compiled_dag_node.py

```diff
@@ -985,6 +987,105 @@ def _add_node(self, node: "ray.dag.DAGNode") -> None:
+            raise ValueError(
+                "NCCL P2P operation is only supported with ClassMethodNode"
+            )
+        elif node.is_cgraph_output_node:
+            raise ValueError(
+                "Outputs cannot be transferred via NCCL because the driver "
+                "cannot participate in the NCCL group"
+            )
+
+        send_actor_handle: "ray.actor.ActorHandle" = node._get_actor_handle()
+        assert send_actor_handle is not None, "Expected an actor handle"
+        send_node = _P2PSendNode(
+            method_args=(node,),
+            other_args_to_resolve={
+                PARENT_CLASS_NODE_KEY: send_actor_handle,
+                P2P_OPERATION_KEY: _P2POperation(),
+                BIND_INDEX_KEY: node._get_bind_index(),
+            },
+        )
+        send_node._type_hint = node.type_hint
+        send_node._original_type_hint = node._original_type_hint
+        node._type_hint = ChannelOutputType()
+        node._original_type_hint = None
+        node_to_p2p_send_node[node] = send_node
+        self._add_node(send_node)
+
+    def _add_nccl_p2p_recv_nodes(
+        self,
+        node: "ray.dag.DAGNode",
+        node_to_p2p_send_node: Dict["ray.dag.DAGNode", "ray.dag.p2p_node._P2PSendNode"],
+    ) -> None:
+        """
+        Add a NCCL P2P recv node to the DAG for each upstream node that requires
+        NCCL send.
+
+        Args:
+            node: A DAG node.
+            p2p_send_dag_nodes: The set of DAG nodes that require NCCL send.
+            node_to_p2p_send_node: A dictionary mapping DAG nodes to their
+                corresponding NCCL P2P send nodes.
+        """
+        from ray.dag import DAGNode, ClassMethodNode, MultiOutputNode
+        from ray.dag.p2p_node import _P2PRecvNode
+
+        new_args = []
+        for arg in node.get_args():
+            if not isinstance(arg, DAGNode) or arg not in node_to_p2p_send_node:
+                new_args.append(arg)
+                continue
+
+            if isinstance(node, MultiOutputNode):
+                raise ValueError(
+                    "Outputs cannot be transferred via NCCL because the driver "
+                    "cannot participate in the NCCL group"
+                )
+            elif not isinstance(node, ClassMethodNode):
+                raise ValueError(
+                    "NCCL P2P operation is only supported with ClassMethodNode"
+                )
+
+            send_node = node_to_p2p_send_node[arg]
+            recv_actor_handle: "ray.actor.ActorHandle" = node._get_actor_handle()
+            assert recv_actor_handle is not None, "Expected an actor handle"
+            recv_node = _P2PRecvNode(
+                method_args=(send_node,),
+                other_args_to_resolve={
+                    PARENT_CLASS_NODE_KEY: recv_actor_handle,
+                    P2P_OPERATION_KEY: send_node.nccl_op,
+                    BIND_INDEX_KEY: node._get_bind_index(),
```

Comment:
Does this mean that the send and recv nodes will get the same bind index as the original DAG node? Will it be a problem that multiple nodes have the same bind index?